### DISCUSSION TITLE
Discrete Ordinates Method in Ferrite?

### DISCUSSION THREAD

**Discussion by warm-considering** [Ferrite-FEM/Ferrite.jl#862] [Q&A] [answered]
I would like to implement an iterative solver using the discrete ordinates method in 2D using Ferrite, does that seem reasonable/plausible? My current thought is to assemble then solve each direction separately (potentially in parallel) with the angular flux for that direction as a single scalar field to solve for. I was thinking of using Graphs.jl to check for cycles in the graph for each direction, then manually deal with the contributions from each direction at the end of each iteration.

Is there a better approach to this? I am inexperienced with implementing FEM solvers so apologies if the answer to this is obvious.

**Comment by termi-official** [comment]
Hi @warm-considering and welcome to the community!

Could you provide more details or a reference with a simple description of the method? I have unfortunately never heard of it. From your description I take that it is it like an alternating-direction implicit (ADI) method?

**Reply by warm-considering** [reply]
Hi, I am unfortunately unfamiliar with the ADI method but I don't believe the methods are related.

The discrete ordinates method is used for radiative transfer problems and for neutron transport problems in reactor physics. I'm interested in the latter.

The idea is that a symmetric angular quadrature is used to discretise the angular dependence of the problem and then angular fluxes are solved for by "sweeping" through the mesh in each discrete direction. My understanding is that if done correctly, determining the correct order of elements for a given sweep should result in a block lower triangular stiffness matrix.

The sweep can be represented as a directed graph in which nodes are mesh elements and edges are faces between elements.

As a reference I found this useful: https://www.osti.gov/biblio/4491151
Though unfortunately it doesn't go into much detail about the discrete ordinates method itself.

Hopefully this is somewhat useful?

**Reply by termi-official** [reply, marked as answer]
Ah, I see. Yes, this is indeed different from ADI.

Generally speaking Ferrite.jl will provide you with low level utilities, like for example the evaluation of quadrature rules over elements and the determination of element neighborhoods. You can interpret a face as a sorted tuple of the attached vertices, which gives you a unique representation for conforming meshes. We also should have all utilities in place for discontinuous solution approximations on master. The last release (v0.3.14) does not include these features yet. We will merge an example to solve a diffusion problem with the symmetric interior penalty method soon (https://github.com/Ferrite-FEM/Ferrite.jl/pull/787) which should give you at least a starting point.

**Reply by warm-considering** [reply]
Ah that's useful to know, thank you. It seems like what I'd like to do should be feasible then. Thanks for your time!
